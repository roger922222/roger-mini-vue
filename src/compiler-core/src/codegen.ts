export function generate(ast) {

  const context = createCodegenContext()

  const { push } = context

  // let code = ''
  // code += 'export '
  push('export ')
  const functionName = 'render'
  const args = ['_ctx', '_cache', '$props', '$setup', '$data', '$options']
  const signature = args.join(', ')
  // code += `function ${functionName}(${signature}){`
  // code += 'return '
  push(`function ${functionName}(${signature}){`)
  push('return ')
  genNode(ast.codegenNode, context)
  // code += `return '${node.content}'` 
  // const node = ast.codegenNode
  // code += '}'
  push('}')
 
  return {
    code: context.code
  }
  
}

function genNode(node, context) {
  const { push } = context
  push(`'${node.content}'`)
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    }
  }
  return context
}