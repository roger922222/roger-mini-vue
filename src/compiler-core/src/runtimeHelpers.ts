export const TO_DISPLAY_STRING = Symbol('toDisplayString')
export const CREATE_ELEMENT_BLOCK = Symbol('createElementBlock')

export const helperMapName = {
  [TO_DISPLAY_STRING]: 'toDisplayString',
  [CREATE_ELEMENT_BLOCK]: 'createElementBlock'
}