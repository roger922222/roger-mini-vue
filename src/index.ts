// mini-vue的出口
export * from './runtime-dom/index'
export * from './reactivity/index'