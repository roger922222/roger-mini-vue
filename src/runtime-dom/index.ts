import { createRenderer } from '../runtime-core/index'
export * from '../runtime-core/index'

function createElement(type) {
  return document.createElement(type)
}

function patchProp(el, key, val) {
  const isOn = (key: string) => /^on[A-Z]/.test(key) 
  if (isOn(key)) {
    const event = key.slice(2).toLowerCase()
    el.addEventListener(event, val)
  } else {
    el.setAttribute(key, val)
  }
}

function insert(el, parent) {
  parent.append(el)
}

const renderer: any = createRenderer({
  createElement, patchProp, insert
})

export function createApp(...args) {
  return renderer.createApp(...args)
}