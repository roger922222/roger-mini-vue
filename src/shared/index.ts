export const extend = Object.assign

export const EMPTY_OBJ = {}

export const isString = value => typeof value === 'string'

export const isObject = val => {
  return val !== null && typeof val === 'object'
}

export const hasChanged = (val, newVal) => {
  return !Object.is(val, newVal)
}

export const hasOwn = (obj, key) => Object.prototype.hasOwnProperty.call(obj, key)

// 串联形式变为驼峰命名的方式
export const camelize = (str: string) => {
  // 第一个参数获取的是-(\w)  第二个参数获取的是括号里面的   replace回调函数的return 值 把第一个参数替换
  return str.replace(/-(\w)/g, (_, c: string) => {
    return c ? c.toUpperCase() : ''
  })
}

//  获取首字母 charAt(0) 转为大写
export const capitalize = (str: string) => {
  return str.charAt(0).toUpperCase() + str.slice(1)
}


// 处理on的行为
export const toHandlerKey = (str: string) => {
  return str ? 'on' + capitalize(str) : ''
}