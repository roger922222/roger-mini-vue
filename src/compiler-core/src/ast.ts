import { CREATE_ELEMENT_BLOCK } from "./runtimeHelpers"

export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_INTERPOLATION,
  ELEMENT,
  TEXT,
  ROOT,
  COMPOUND_EXPRESSION,
}

export function createVNodeCall(context, tag, props, children) {
  context.helper(CREATE_ELEMENT_BLOCK)
  return {
    type: NodeTypes.ELEMENT,
    tag,
    props,
    children
  }
}