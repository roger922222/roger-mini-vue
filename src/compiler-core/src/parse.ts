import { NodeTypes } from "./ast"

export function baseParse(content: string) {

  const context = createParseContext(content)

  return createRoot(parseChildren(context))
}

function parseInterpolation(context) {

  // 接受的 {{message}}

  const openDelimiter = '{{'
  const closeDelimiter = '}}'

  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)

  // context.source = context.source.slice(openDelimiter.length)
  advanceBy(context, openDelimiter.length)

  const rawContentLength = closeIndex - openDelimiter.length

  const rawContent = context.source.slice(0, rawContentLength)
  const content = rawContent.trim()
 
  // context.source = context.source.slice(rawContentLength + closeDelimiter.length)
  advanceBy(context, rawContentLength + closeDelimiter.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_INTERPOLATION,
      content: content
    }
  }
}

function parseChildren(context) {
  const nodes: any = []
  let node
  if (context.source.startsWith('{{')) {
    node = parseInterpolation(context)
  }
  nodes.push(node)
  return nodes
}

function createRoot(children) {
  return {
    children
  }
}

function createParseContext(content: string) {
  return {
    source: content
  }
}

function advanceBy(context: any, length: number) {
  context.source = context.source.slice(length)
}