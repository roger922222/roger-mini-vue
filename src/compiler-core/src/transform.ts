import { NodeTypes } from "./ast"

export function transform(root, options) {
  const context = createTransformContext(root, options)

  // 1. 遍历 --- 深度优先搜索
  traverseNode(root, context)

  // 2. 修改text --- content
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || []
  }

  return context
}

function traverseNode(node, context) {
  
  // if (node.type === NodeTypes.TEXT) {
  //   node.content = node.content + ' mini-vue'
  // }

  const nodeTransforms = context.nodeTransforms
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    transform(node)
  }

  traverseNodeChildren(node, context)

}

function traverseNodeChildren(node, context) {
  const children = node.children

  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i]

      traverseNode(node, context)
    }
  }  
}