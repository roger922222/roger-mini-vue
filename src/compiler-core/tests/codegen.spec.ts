import { generate } from "../src/codegen"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"

describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)

    // 快照测试 给当前的code，拍照，后续进行对比
    // 1. 抓bug
    // 2. 主动的更新快照
    expect(code).toMatchSnapshot()
  })
})