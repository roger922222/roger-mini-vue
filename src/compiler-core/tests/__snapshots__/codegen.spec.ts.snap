// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"import { createElementBlock as _createElementBlock } from "vue"
export function render(_ctx, _cache, $props, $setup, $data, $options){return _createElementBlock('div')}"
`;

exports[`codegen element-string-interpolation 1`] = `
"import { toDisplayString as _toDisplayString, createElementBlock as _createElementBlock } from "vue"
export function render(_ctx, _cache, $props, $setup, $data, $options){return _createElementBlock('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`codegen interpolation 1`] = `
"import { toDisplayString as _toDisplayString } from "vue"
export function render(_ctx, _cache, $props, $setup, $data, $options){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
export function render(_ctx, _cache, $props, $setup, $data, $options){return 'hi'}"
`;
